src/execution_plan/ops/op_merge.c

```diff
@@ -15,193 +14,289 @@
-		if(!schema) schema = GraphContext_AddSchema(op->gc, edge_blueprint->reltypes[0], SCHEMA_EDGE);
+//------------------------------------------------------------------------------
+// Merge logic
+//------------------------------------------------------------------------------
+static inline Record _pullFromStream(OpBase *branch) {
+	return OpBase_Consume(branch);
+}
 
-		// Node are already created, get them from record.
-		EntityID srcId = ENTITY_GET_ID(Record_GetNode(r, edge_ctx->src_idx));
-		EntityID destId = ENTITY_GET_ID(Record_GetNode(r, edge_ctx->dest_idx));
+OpBase *NewMergeOp(const ExecutionPlan *plan, EntityUpdateEvalCtx *on_match,
+				   ResultSetStatistics *stats) {
+	/* Merge is an operator with two or three children.They will be created outside of here,
+	 * as with other multi-stream operators (see CartesianProduct and ValueHashJoin). */
+	OpMerge *op = calloc(1, sizeof(OpMerge));
+	op->stats = stats;
+	op->on_match = on_match;
 
-		assert(Graph_ConnectNodes(op->gc->g, srcId, destId, schema->id, created_edge));
+	// Set our Op operations
+	OpBase_Init((OpBase *)op, OPType_MERGE, "Merge", NULL, MergeConsume, NULL, NULL, MergeFree, plan);
 
-		// Convert properties and add to newly-created node.
-		PropertyMap *map = edge_ctx->properties;
-		if(map) _AddProperties(op, r, (GraphEntity *)created_edge, map);
+	if(op->on_match) {
+		// If we have ON MATCH directives, set the appropriate record IDs of entities to be updated.
+		uint on_match_count = array_len(op->on_match);
+		for(uint i = 0; i < on_match_count; i ++) {
+			op->on_match[i].record_idx = OpBase_Modifies((OpBase *)op, op->on_match[i].alias);
+		}
 	}
 
-	if(op->stats) op->stats->relationships_created += edge_count;
+	return (OpBase *)op;
 }
 
-static void _CreateEntities(OpMerge *op, Record r) {
-	// Track the inherited Record and the newly-allocated Record so that they may be freed if execution fails.
-	OpBase_AddVolatileRecord((OpBase *)op, r);
+static Record _handoff(OpMerge *op) {
+	Record r = NULL;
+	if(array_len(op->output_records)) r = array_pop(op->output_records);
+	return r;
+}
 
-	// Lock everything.
-	Graph_AcquireWriteLock(op->gc->g);
+static Record MergeConsume(OpBase *opBase) {
+	OpMerge *op = (OpMerge *)opBase;
 
-	// Commit query graph and set resultset statistics.
-	_CommitNodes(op, r);
-	_CommitEdges(op, r);
+	// Return mode, all data was consumed.
+	if(op->output_records) return _handoff(op);
 
-	// Release lock.
-	Graph_ReleaseLock(op->gc->g);
+	// Consume mode.
+	op->output_records = array_new(Record, 32);
 
-	OpBase_RemoveVolatileRecords((OpBase *)op); // No exceptions encountered, Records are not dangling.
+	// If we have a bound variable stream, pull from it and store records until depleted.
+	if(op->bound_variable_stream) {
+		Record input_record;
+		while((input_record = _pullFromStream(op->bound_variable_stream))) {
+			op->input_records = array_append(op->input_records, input_record);
+		}
+	}
```

Comment:
This is an interesting point! This can be rewritten such that this loop is within the `reading_matches` loop at 147, but I think we have a broader problem using LIMIT with any eager op / op with a handoff stage. For example:
`UNWIND [1,2,3] as val CREATE (b:b {v: val}) RETURN b LIMIT 1`
For us this generates 3 nodes; for Neo it creates 1.

In general I think that LIMIT has marginal utility for users in data-modifying contexts, but we still need to normalize this behavior.